### ISSUE TITLE
How to prune a custom tensor? The tensor is a recursive variable and is initialized with tf.zeros.

### ISSUE THREAD

**Issue by starsky68** [tensorflow/model-optimization#875]
Prior to filing: check that this should be a bug instead of a feature request. Everything supported, including the compatible versions of TensorFlow, is listed in the overview page of each technique. For example, the overview page of quantization-aware training is [here](https://www.tensorflow.org/model_optimization/guide/quantization/training). An issue for anything not supported should be a feature request.

**Describe the bug**
How to prune a custom tensor? The tensor is a recursive variable and is initialized with tf.zeros.

**System information**

TensorFlow version (installed from source or binary):

TensorFlow Model Optimization version (installed from source or binary):

Python version:

**Describe the expected behavior**

**Describe the current behavior**
How to prune "b"
**Code to reproduce the issue**
class PruningLayer(tf.keras.layers.Layer, tfmot.sparsity.keras.PrunableLayer):

    def __init__(self, n, d):
        super(PruningLayer, self).__init__()

        self.n = n
        self.d = d
    def build(self, input_shape):
        self.weight = self.add_weight("weight", shape=[1, input_shape[1],
                                                       self.n,
                                                       self.d,
                                                       input_shape[2]
                                                       ],
                                      initializer="random_normal",
                                      trainable=True)

    def call(self, x):
        x = tf.expand_dims(x, axis=2)
        b = self.rout(x)
        c = tf.nn.softmax(b, axis=2)
        s = tf.reduce_sum(tf.multiply(u, c), axis=1)
        return s

    def get_prunable_weights(self):
        return [self.weight]

    def rout(self, x):
        input_shape = tf.shape(x)
        # initialize b to zero
        b = tf.zeros((input_shape[0], input_shape[1], self.num_capsule, 1))

        for _ in range(3):
            c = tf.nn.softmax(b, axis=2)
            b = b + tf.reduce_sum(tf.multiply(x, c), axis=-1, keepdims=True)
        return b

**Screenshots**

**Additional context**

**Comment by fredrec** [comment]
Hi @starsky68,

Can you please provide more context ?
In particular:
- What is `self.weight` for ? It looks like it is never used in the layer.
- What prevents `b` to be a class member ? You could then return it in `get_prunable_weights(self)`

**Comment by starsky68** [comment]
> 嗨@starsky68，
> 
> 你能提供更多的背景吗？ 特别是：
> 
> * 是`self.weight`为了什么？看起来它从未在图层中使用过。
> * 什么阻止`b`成为班级成员？然后你可以把它退回`get_prunable_weights(self)`

I modified the above sample code again. Self. Weight can be obtained through get_ prunable_ Weights returns, but I don't know if this B is returned to get_ prunable_ Weights, hope to get help

**Comment by starsky68** [comment]
> Hi @starsky68,
> 
> Can you please provide more context ? In particular:
> 
> * What is `self.weight` for ? It looks like it is never used in the layer.
> * What prevents `b` to be a class member ? You could then return it in `get_prunable_weights(self)`

I modified the above sample code again. Self. Weight can be obtained through get_ prunable_ Weights returns, but I don't know if this B is returned to get_ prunable_ Weights, hope to get help

**Comment by starsky68** [comment]
> Hi @starsky68,
> 
> Can you please provide more context ? In particular:
> 
> * What is `self.weight` for ? It looks like it is never used in the layer.
> * What prevents `b` to be a class member ? You could then return it in `get_prunable_weights(self)`

When I use TF1, I can directly use its pruning interface apply_ Mask operates on the tensor, but the current interface seems to have changed after TF2. Such operations are no longer supported
